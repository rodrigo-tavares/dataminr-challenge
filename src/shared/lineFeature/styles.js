import styled from "styled-components";

export const ContainerLine = styled.div`
  margin-bottom: 5px;
  position: relative;
  ${({ isGroupLine }) => (isGroupLine ? `flex: 1;` : `flex: 0 0 32.5%;`)}
  background-color: #18191d;
  border: 2px solid #1b1c20;
  border-radius: 5px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`;

export const ActionButtons = styled.div`
  position: relative;
  display: flex;
  width: 29%;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;

  ${({ isGroupLine }) => isGroupLine && `padding-right: 20px;`}
`;

export const Line = styled.div`
  height: 60px;
  width: 70%;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const LineLabel = styled.div`
  width: 100%;
  font-size: 14px;
  text-transform: uppercase;
  position: relative;
  padding-left: 45px;
  font-weight: 700;

  @media only screen and (min-width: 720px) {
    font-size: 16px;
  }

  &:before {
    content: "";
    position: absolute;
    width: 24px;
    height: 24px;
    background-color: #707070;
    border-radius: 5px;
    left: 10px;
    top: -4px;
  }
`;
