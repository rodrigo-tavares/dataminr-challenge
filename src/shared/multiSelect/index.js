import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { faChevronDown, faChevronUp } from "@fortawesome/free-solid-svg-icons";
import {
  SelectContainer,
  SelectHeader,
  SelectItem,
  SelectOptions,
} from "./styles";

export const MultiSelect = ({ values, selectedValue }) => {
  const [isOpen, setOpen] = useState(false);
  const [selectedItem, setSelectedItem] = useState(selectedValue);

  const toggleSelect = () => setOpen(!isOpen);

  const handleItemClick = (value) => {
    setSelectedItem(value);
    setOpen(!isOpen);
  };

  return (
    <SelectContainer>
      <SelectHeader onClick={toggleSelect}>
        {selectedItem
          ? values.find((item) => item.value === selectedItem).label
          : null}
        <FontAwesomeIcon icon={isOpen ? faChevronUp : faChevronDown} />
      </SelectHeader>
      <SelectOptions isOpen={isOpen}>
        {values.map((item, key) => (
            <SelectItem onClick={() => handleItemClick(item.value)} key={key}>
              {item.label}
            </SelectItem>
          ))}
      </SelectOptions>
    </SelectContainer>
  );
};
