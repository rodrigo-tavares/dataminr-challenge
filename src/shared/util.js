export const NormalizeOptions = (maxValue, frequency) => {
  return Array.from(Array(maxValue + 1).keys()).map((item) => ({
    label: item,
    value: item,
  })).filter(item => item.value % frequency === 0 && item.value !== 0);
};
