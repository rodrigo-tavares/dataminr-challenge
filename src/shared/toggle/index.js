import { ToggleButton } from "./styles"

export const Toggle = ({value, onClick}) => {
    return <ToggleButton isEnabled={value} onClick={onClick}/>
}