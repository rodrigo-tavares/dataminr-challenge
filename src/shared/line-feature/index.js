import { useState } from "react";
import { Toggle } from "../toggle";
import { ContainerLine, Line, LineLabel } from "./styles";

export const LineFeature = ({ name, value }) => {
  const [enabled, setEnabled] = useState(value);

  const onChangeToggle = () => {
    setEnabled(!enabled);
  };

  return (
    <ContainerLine>
      <Line>
        <LineLabel>{name}</LineLabel>
      </Line>

      <Toggle value={enabled} onClick={onChangeToggle} />
    </ContainerLine>
  );
};
