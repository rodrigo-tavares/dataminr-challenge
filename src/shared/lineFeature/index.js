import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { faChevronDown, faChevronUp } from "@fortawesome/free-solid-svg-icons";

import { Toggle } from "../toggle";
import { ActionButtons, ContainerLine, Line, LineLabel } from "./styles";
import { MultiSelect } from "../multiSelect";
import { NormalizeOptions } from "../util";

export const LineFeature = ({
  name,
  value,
  onChange,
  hasChildren,
  isGroupLine = false,
  maxValue = 0,
  selectedValue = null,
  frequency = 0,
}) => {
  const [toggleValue, setToggleValue] = useState(value);

  const onChangeToggle = () => {
    setToggleValue(!toggleValue);
    onChange(!toggleValue);
  };

  const selectOptions = NormalizeOptions(maxValue, frequency);

  return (
    <ContainerLine isGroupLine={isGroupLine}>
      <Line>
        <LineLabel>{name}</LineLabel>
      </Line>
      <ActionButtons isGroupLine={isGroupLine && hasChildren}>
        {selectOptions && selectOptions.length > 0 && (
          <MultiSelect
            key={name}
            values={selectOptions}
            selectedValue={selectedValue}
          />
        )}
        <Toggle value={toggleValue} onClick={onChangeToggle} />
        {isGroupLine && hasChildren && (
          <FontAwesomeIcon icon={toggleValue ? faChevronUp : faChevronDown} />
        )}
      </ActionButtons>
    </ContainerLine>
  );
};
