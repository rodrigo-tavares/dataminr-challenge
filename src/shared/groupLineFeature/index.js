// import { ContainerLine } from "./../lineFeature/styles";
import { useState } from "react";
import { LineFeature } from "../lineFeature";
import { ChildrenContainerGroup, ContainerGroup } from "./styles";

export const GroupLineFeature = ({ childs }) => {
  const [childrenEnabled, setChildrenEnabled] = useState({});

  const validChildren = (children) => children && children.length > 0;

  const renderChildren = (children) => {
    return (
      <ChildrenContainerGroup>
        {children.map((child, key) => (
          <LineFeature
            key={key}
            name={child.title}
            value={child.value}
            isGroupLine={true}
            hasChildren={false}
            onChange={() => {}}
            maxValue={child.maxValue}
            selectedValue={child.selectedValue}
            frequency={child.frequency}
          />
        ))}
      </ChildrenContainerGroup>
    );
  };

  const onToggleChange = (key, value) => {
    setChildrenEnabled({ [key]: value });
  };

  return (
    <ContainerGroup>
      {childs.map((child, key) => (
        <>
          <LineFeature
            key={key}
            name={child.title}
            value={child.value}
            isGroupLine={true}
            hasChildren={validChildren(child.children)}
            onChange={(value) => onToggleChange(key, value)}
            maxValue={child.maxValue}
            selectedValue={child.selectedValue}
            frequency={child.frequency}
          />
          {validChildren(child.children) &&
            childrenEnabled[key] &&
            renderChildren(child.children)}
        </>
      ))}
    </ContainerGroup>
  );
};
