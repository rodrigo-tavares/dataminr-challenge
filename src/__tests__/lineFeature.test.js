import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import { LineFeature } from "../shared/lineFeature";

const props = {
  name: "test",
  value: true,
  onChange: jest.fn(),
  hasChildren: true,
  isGroupLine: false,
  maxValue: 100,
  selectedValue: 20,
  frequency: 10,
};
describe("Line Feature component", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<LineFeature {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should receive a toggle value false", () => {
    const customProps = {
        value: false,
      },
      view = mount(<LineFeature {...customProps} />);

    expect(view.prop("value")).toEqual(false);
  });

  it("should click toggle button", () => {
    const onChange = jest.fn(),
      customProps = {
        value: true,
        onChange,
      },
      view = mount(<LineFeature {...customProps} />).find("Toggle");
    view.simulate("click");
    expect(onChange).toHaveBeenCalled();
  });
});
