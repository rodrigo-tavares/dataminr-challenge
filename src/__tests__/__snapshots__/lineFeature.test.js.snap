// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line Feature component renders correctly 1`] = `
<div
  className="sc-gsDKAQ gagpPD"
>
  <div
    className="sc-hKwDye hza-dq"
  >
    <div
      className="sc-eCImPb flEqpm"
    >
      test
    </div>
  </div>
  <div
    className="sc-dkPtRN hJqBHD"
  >
    <div
      className="sc-jRQBWg cjCSSX"
    >
      <div
        className="sc-gKclnd Mwisz"
        onClick={[Function]}
      >
        20
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-w-14 "
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <div
        className="sc-iCfMLu joRmOP"
      >
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          10
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          20
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          30
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          40
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          50
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          60
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          70
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          80
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          90
        </div>
        <div
          className="sc-furwcr UmGsm"
          onClick={[Function]}
        >
          100
        </div>
      </div>
    </div>
    <div
      className="sc-bdvvtL bbFHQR"
      onClick={[Function]}
    />
  </div>
</div>
`;
